File: 01.ListasEnlazadas/lista.c
#include "lista.h"
#include <stdlib.h>
#include <string.h>

Lista* crearLista(char propietario[]){
    Lista* listita = (Lista*) malloc(sizeof (Lista));
    printf("\n Creando lista...");

    strncpy(listita->propietario, propietario, 50);
    listita->longitud = 0;
    printf(" Lista creada. \n");
    return listita;
}

Nodo* CrearNodo(Libro* librito){
    Nodo* nodo = (Nodo*) malloc(sizeof (Nodo));
    strncpy(nodo->libro.titulo, librito->titulo, 50);
    strncpy(nodo->libro.autor, librito->autor, 50);
    strncpy(nodo->libro.isbn, librito->isbn, 13);
    nodo->siguiente = 0;
    return nodo;
}

void DestruirNodo(Nodo* nodo){
    free(nodo);
}

void InsertarPrincipio(Lista* lista, Libro* libro){
    Nodo* nodo = CrearNodo(libro);
    nodo->siguiente = lista->cabeza;
    lista->cabeza = nodo;
    lista->longitud++; 
}

void InsertarFinal(Lista* lista, Libro* libro){
    Nodo* nodo = CrearNodo(libro);
    Nodo* puntero = lista->cabeza;

    if(lista->cabeza == 0){
        lista->cabeza = nodo;
    }else{
        Nodo* puntero = lista->cabeza;
        while (puntero->siguiente){
            puntero = puntero->siguiente;
        } 
        puntero->siguiente = nodo;
    }
    lista->longitud++;
}

void InsertarLibro(int posicion, Lista* lista, Libro* libro){
    Nodo* nodo = CrearNodo(libro);

    if(lista->cabeza == 0){
        lista->cabeza = nodo;
    } else {
        Nodo* puntero = lista->cabeza;
        int posicion = 0;

         while (posicion < posicion && puntero->siguiente)
        {
            puntero = puntero->siguiente;
            posicion++;
        }
        nodo->siguiente = puntero->siguiente;
        puntero->siguiente = nodo;
    } 
    lista->longitud++;
}

Libro* Obtener(int n, Lista* lista){
    if(lista->cabeza == 0){
        return 0;
    }else{
        Nodo* puntero = lista->cabeza;
        int posicion = 0;

        while (posicion < n && puntero->siguiente)
        {
            puntero = puntero->siguiente;
            posicion++;
        }

        if(posicion != n){
            return 0;
        } else
        {
            return &puntero->libro;
        }
    }
}

int Contar(Lista* lista){
    return lista->longitud;
}

int estaVacia(Lista* lista){
    return lista->cabeza == 0;
}

void EliminarPrimer(Lista* lista){

    if(lista->cabeza){
        Nodo* eliminar = lista->cabeza;
        lista->cabeza = lista->cabeza->siguiente;
        DestruirNodo(eliminar);
        lista->longitud--;
    }
}

void EliminarUltimo(Lista* lista){
    if(lista->cabeza){

        if(lista->cabeza->siguiente){
            Nodo* puntero = lista->cabeza;

            while (puntero->siguiente->siguiente)
            {
                puntero = puntero->siguiente;
                /* code */
            }
            Nodo* eliminar = puntero->siguiente;
            puntero->siguiente = 0;
            DestruirNodo(eliminar);
            lista->longitud--;

        } else
        {
            Nodo* eliminar = lista->cabeza;
            lista->cabeza = 0;
            DestruirNodo(eliminar);
            lista->longitud--;
        }   
    }
}

void Eliminar(int posicionAEliminar,Lista* lista){
    if(lista->cabeza) {
        if(posicionAEliminar==0){
            Nodo* eliminar = lista->cabeza;
            lista->cabeza = lista->cabeza->siguiente;
            DestruirNodo(eliminar);
            lista->longitud--;
        } else if(posicionAEliminar < lista->longitud){
            Nodo* puntero = lista->cabeza;
            int posicion = 0;
            while (posicion < (posicionAEliminar-1)) {
                puntero = puntero->siguiente;
                posicion++;
            }
            Nodo* eliminar = puntero->siguiente;
            puntero->siguiente = eliminar->siguiente;
            DestruirNodo(eliminar);
            lista->longitud--;
        }
    }
}

File: java/01.ListasEnlazadas/Lista.java
public class Lista {
    private Nodo cabeza;
    private String propietario;
    private int longitud = 0;

    private class Nodo {

        public Libro libro;
        public Nodo siguiente = null;

        public Nodo(Libro libro) {
            this.libro = libro;
        }
    }

    public Lista(String propietario) {
        this.propietario = propietario;
    }

    /**
     * @return the propietario
     */
    public String getPropietario() {
        return propietario;
    }

    /**
     * @param propietario the propietario to set
     */
    public void setPropietario(String propietario) {
        this.propietario = propietario;
    }

    public void insertarPrincipio(Libro libro) {
        Nodo nodo = new Nodo(libro);
        nodo.siguiente = cabeza;
        cabeza = nodo;
        longitud++;
    }

    public void insertarFinal(Libro libro) {
        Nodo nodo = new Nodo(libro);

        if (cabeza == null) {
            cabeza = nodo;
        } else {
            Nodo puntero = cabeza;
            while (puntero.siguiente != null) {
                puntero = puntero.siguiente;
            }
            puntero.siguiente = nodo;
        }
        longitud++;
    }

    public void insertarLibro(int posicion, Libro libro) {
        Nodo nodo = new Nodo(libro);

        if (cabeza == null) {
            cabeza = nodo;
        } else {
            Nodo puntero = cabeza;
            int contador = 0;

            while (contador < posicion && puntero.siguiente != null) {
                puntero = puntero.siguiente;
                contador++;
            }

            nodo.siguiente = puntero.siguiente;
            puntero.siguiente = nodo;
        }
        longitud++;
    }

    public Libro getLibro(int n) {
        if (cabeza == null) {
            return null;
        } else {
            Nodo puntero = cabeza;
            int contador = 0;

            while (contador < n && puntero.siguiente != null) {
                puntero = puntero.siguiente;
                contador++;
            }

            if (contador != n) {
                return null;
            } else {
                return puntero.libro;
            }
        }
    }

    public int contar() {
        return longitud;
    }

    public boolean estaVacia() {
        return cabeza == null;
    }

    public void eliminarPrincipio() {
        if (cabeza != null) {
            Nodo primero = cabeza;
            cabeza = cabeza.siguiente;
            primero.siguiente = null;
            longitud--;
        }
    }

    public void eliminarUltimo() {
        if (cabeza != null) {

            if (cabeza.siguiente.siguiente == null) {
                cabeza = null;
            } else {
                Nodo puntero = cabeza;
                while (puntero.siguiente.siguiente != null) {
                    puntero = puntero.siguiente;
                }
                puntero.siguiente = null;
            }
            longitud--;
        }
    }

    public void eliminarLibro(int n) {
        if (cabeza != null) {

            if (n == 0) {
                Nodo primero = cabeza;
                cabeza = cabeza.siguiente;
                primero.siguiente = null;
            } else if (n < longitud) {
                Nodo puntero = cabeza;
                int contador = 0;
                while (contador < (n - 1)) {
                    puntero = puntero.siguiente;
                    contador++;
                }

                Nodo temp = puntero.siguiente;
                puntero.siguiente = temp.siguiente;
                temp.siguiente = null;
            }

            longitud--;
        }
    }

}



